crypto/x509/x509_set.c

```diff
@@ -238,3 +239,67 @@ const X509_ALGOR *X509_get0_tbs_sigalg(const X509 *x) {
 X509_PUBKEY *X509_get_X509_PUBKEY(const X509 *x509) {
   return x509->cert_info->key;
 }
+
+static int X509_SIG_INFO_get(const X509_SIG_INFO *sig_info, int *digest_nid,
+                             int *pubkey_nid, int *sec_bits, uint32_t *flags) {
+  if (digest_nid != NULL) {
+    *digest_nid = sig_info->digest_nid;
+  }
+  if (pubkey_nid != NULL) {
+    *pubkey_nid = sig_info->pubkey_nid;
+  }
+  if (sec_bits != NULL) {
+    *sec_bits = sig_info->sec_bits;
+  }
+  if (flags != NULL) {
+    *flags = sig_info->flags;
+  }
+  return (sig_info->flags & X509_SIG_INFO_VALID) != 0;
+}
+
+int X509_get_signature_info(X509 *x509, int *digest_nid, int *pubkey_nid,
+                            int *sec_bits, uint32_t *flags) {
+  if (!X509_check_purpose(x509, -1, -1)) {
```

Comment:
It seems like the intent of this is to just trigger `x509v3_cache_extensions`, as it will bail out of the function afterwards due to `id == -1`.  Would it be clearer to just call the cache function?

Reply:
Ahh nice catch, I was just following the original openssl implementation, but that makes sense. I'll also add [this comment](https://github.com/openssl/openssl/blame/master/crypto/x509/v3_purp.c#L83C3-L84) from OpenSSL to `X509_check_purpose`

Reply:
It seems like Postgres is also [dependent](https://github.com/postgres/postgres/blob/0075d78947e3800c5a807f48fd901f16db91101b/src/backend/libpq/be-secure-openssl.c#L1470-L1479) on this function and the new behavior is causing the postgres CI to fail..
I'm guessing that `X509_get_signature_info` is just a information retrieval function, so the cert is allowed to "invalid extensions".. which explains why [OpenSSL's implementation](https://github.com/openssl/openssl/blob/master/crypto/x509/x509_set.c#L204-L209) didn't check the return value. It's a bit of a pain, but I'll have to change this to be less error prone

Reply:
Nvm, the postgres CI was failing since the initial change. I'll have to figure out what's going on here

Reply:
I see all CI checks have passed. What ended up being the issue with the postgres checks?

Reply:
I forgot to mention I cut `CryptoAlg-2393` to track PSS support in libssl. I'll add that in the commit/PR description.